### PR TITLE
Blip4

### PR DESCRIPTION
**Pull request by ipatka**
Breaking change

Change old Merkle tree format to 'legacy' methods

2 new Merkle tree methods. getSignedMerkleTreeComponents and getSignedBatchMerkleTreeComponents

1. Subject provides data to attester
2. Attester calls `getSignedMerkleTreeComponents` and passes root hash back to subject to sign
3. If using the new batch tree structure, attester calls `getSignedBatchMerkleTreeComponents` to incorporate subjectSig into layer2hash

compare with spec here:
https://github.com/hellobloom/specs/blob/fc631eab2c1e1a30ad77b664080114cc6ad3cab3/attestation-data/Bloom-Merkle-Tree-22facf0b-bedb-4b45-bb7d-edcd57213eb0.md

### CODE REVIEW

File: src/HashingLogic.ts [hunk 1]
```diff
@@ -69,6 +73,16 @@ export interface IRevocationLinks {
...
+  /**
+   * RFC3339 timestamp of when the claim was issued
+   * https://tools.ietf.org/html/rfc3339
+   */
+  issuanceDate: string
+  /**
+   * RFC3339 timestamp of when the claim should expire
+   * https://tools.ietf.org/html/rfc3339
+   */
+  expirationDate: string
```
Comment (edhedges) [inline src/HashingLogic.ts]: These are just ISO-8601 right?

File: src/HashingLogic.ts [hunk 2]
```diff
@@ -93,23 +107,64 @@ export interface IAttestation {
...
-export interface IDataNode {
-  attestationNode: IAttestationNode
-  signedAttestation: string // Root hash of Attestation tree signed by attester
+export interface ISignedClaimNode {
+  claimNode: IIssuedClaimNode
+  attesterSig: string // Root hash of claim tree signed by attester
+}
+
+export interface IBloomBatchMerkleTreeComponents {
+  batchLayer2Hash: string // Hash of attester sig and subject sig
```
Comment (edhedges) [inline src/HashingLogic.ts]: Are we still going to do on-chain attestations?

File: src/HashingLogic.ts (hunk 1 above)
Comment (ipatka) [reply src/HashingLogic.ts]: it's iso-8601 + a couple more constraints. it's required in the verifiable claims spec draft from w3c

File: src/HashingLogic.ts (hunk 2 above)
Comment (ipatka) [reply src/HashingLogic.ts]: This data structure should be agnostic to that. It's usable without putting it on chain but someone could put it on chain if their use case requires it

**Comment by ipatka** [comment]
The migration path is to: 
1. update attestation kit to support legacy and batch types based on versioned api endpoints for bloom web to use
2. update bloom web to use batch types

**Pull request merged**